core/src/storage2/collections/binary_heap/group.rs

```diff
@@ -0,0 +1,213 @@
+/// contain one element (the root element).
+pub(crate) const COUNT: u32 = 2;
+
+/// Returns the index of the group in which the `n`-th element is stored.
+pub(crate) fn get_group_index(n: u32) -> u32 {
+    match n {
+        0 => 0,
+        _ => {
+            // The first group only ever contains the root element:
+            // `[Some(root), None]`. So when calculating indices we
+            // need to account for the items which have been left
+            // empty in the first group.
+            let padding = COUNT - 1;
+            (n + padding) / COUNT
+        }
+    }
+}
+
+/// Returns the in-group index of the `n`-th element.
+/// This refers to the index which the element has within a group.
+///
+/// For example, the element `3` is found at in-group index `0`
+/// (within the group at index `2`).
+pub(crate) fn get_ingroup_index(n: u32) -> Ingroup {
+    let group = get_group_index(n);
+    match (group, n) {
+        (0, 0) => Ingroup::Left,
+        (0, _) => panic!("first group contains only the root element"),
+        (_, _) => {
+            let ingroup_index = (n - 1) % COUNT;
+            match ingroup_index {
+                0 => Ingroup::Left,
+                1 => Ingroup::Right,
+                _ => {
+                    unreachable!(
+                        "COUNT is 2, following the modulo op index must be 0 or 1"
+                    )
+                }
+            }
+        }
+    }
+}
+
+impl<T> Group<T>
+where
+    T: PackedLayout + Ord,
+{
+    /// Returns a shared reference to the element at `index`.
+    pub fn as_ref(&self, index: u32) -> Option<&T> {
+        let ingroup_index = get_ingroup_index(index);
+        match ingroup_index {
+            Ingroup::Left => self.0.as_ref(),
+            Ingroup::Right => self.1.as_ref(),
+        }
+    }
+
+    /// Returns an exclusive reference to the element at `index`.
+    pub fn get_mut(&mut self, index: u32) -> &mut Option<T> {
+        let ingroup_index = get_ingroup_index(index);
+        match ingroup_index {
+            Ingroup::Left => &mut self.0,
+            Ingroup::Right => &mut self.1,
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+const _: () = {
+    use crate::storage2::traits::StorageLayout;
+    use ink_metadata::layout2::{
+        CellLayout,
+        FieldLayout,
+        Layout,
+        LayoutKey,
+        StructLayout,
+    };
+
+    impl<T> StorageLayout for Group<T>
+    where
+        T: PackedLayout + Ord + TypeInfo + 'static,
+    {
+        fn layout(key_ptr: &mut KeyPtr) -> Layout {
+            StructLayout::new(vec![
+                FieldLayout::new(
+                    None,
+                    CellLayout::new::<i32>(LayoutKey::from(key_ptr.advance_by(1))),
+                ),
+                FieldLayout::new(
+                    None,
+                    CellLayout::new::<i64>(LayoutKey::from(key_ptr.advance_by(1))),
+                ),
+            ])
```

Comment:
This looks wrong. Group signature is `(pub Option<T>, pub Option<T>)`

Reply:
From what I saw this is an `AlwaysPacked` data structure that always pushes and pulls to and from storage in a packed representation using only one cell for the whole data structure. Note that this is the whole point behind this optimization PR. In that vein @ascjones is right because this implementation would yield metadata that indicates that this struct storage its child nodes at 2 different storage locations.

Also I find it strange that is uses `i32` and `i64` for the `CellLayout`s.

Reply:
Inspiration from `Stash` `Header` type:
```rust
impl StorageLayout for Header {
    fn layout(key_ptr: &mut KeyPtr) -> Layout {
        Layout::Cell(CellLayout::new::<Header>(LayoutKey::from(
            key_ptr.advance_by(1),
        )))
    }
}
```